### PR TITLE
Add syncing support for mock interface

### PR DESCRIPTION
**Pull request by altonen**
Add request-response support for the mock interface and rewrite all syncing-related tests to be executed both with mock interface and libp2p.

### CODE REVIEW

File: p2p/src/net/mock/types.rs
Comment (iljakuklic) [inline p2p/src/net/mock/types.rs:112]: This will probably expose some operations that don't make sense on IDs. How many u64 operations are actually nneded, besides comparisons which are auto-derived anyway? Possibly just exposing required operations one by one would not be that much more code.

File: p2p/src/net/mock/request_manager.rs
Comment (iljakuklic) [inline p2p/src/net/mock/request_manager.rs:89]: Since these IDs are u64 (correct me if I'm wrong), how about having 2<spu>32</sup> zones with 2<sup>32</sup> IDs each to support the occasional odd test that needs many requests. Also is the possibility of one peers request IDs overflowing into the next zone checked somewhere so it's detected when it happens?

**Review by iljakuklic — APPROVED** [review]

File: p2p/src/net/mock/request_manager.rs
Comment (altonen) [reply p2p/src/net/mock/request_manager.rs:89]: I'm not entirely opposed to changing the allocation scheme to what you're proposing but I find it more likely that we'll have a test that needs more than `232` zones rather than a test that has `2^32` pending requests. The request ID's are most likely going to require some work either when request timeouts are implemented. I'd prefer doing that work later as it's not in the same rush as pubsub or fixing connectivity issues.

I'm not sure if it's possible for a peer's IDs to overflow into another peer's zone since the size of the zone is capped at `u16::MAX` and zones starts at a multiple of `0x10000`. There is currently one test checking that the allocation/overflow doesn't break at least for the normal case.

Comment (iljakuklic) [reply p2p/src/net/mock/request_manager.rs:89]: That's fair

**Review by stanislav-tkach — APPROVED** [review]

File: p2p/src/net/mock/types.rs
Comment (altonen) [reply p2p/src/net/mock/types.rs:111]: Done

Also simplified the request ID allocation to just use one common counter for all outbound requests and increment it once for each request. No need to deal with peer-specific zones, it was part of previous design.

**Review by iljakuklic — APPROVED** [review]

**Review by TheQuantumPhysicist — APPROVED** [review]

**Review by TheQuantumPhysicist — COMMENTED** [review]
Any reason why this isn't getting merged?

**Comment by altonen** [comment]
I think somebody resrun the tests but last time everything was broken so I was hesitant about merging this.

**Review by muursh — APPROVED** [review]

**Comment by altonen** [comment]
Once the CI succeeds, I think I'll merge this unless there are objections. I can continue debugging the functional tests in some other branch

**Comment by TheQuantumPhysicist** [comment]
> Once the CI succeeds, I think I'll merge this unless there are objections. I can continue debugging the functional tests in some other branch

Go for it. I hope we won't regret this.

**Pull request merged**
